### PR TITLE
LP Presolving - GSOC Project.  

### PR DESCRIPTION
**Pull request by ramcha24**
This is first look at presolve routines implemented at MPB layer. 
Note - 
It currently doesn't support two-sided interval constraints. 
Some presolve routines like forcing constraints and column singletons will be implemented as I continue working on this past the GSOC time-period. 

The code is commented and continually modified for better clarity.  There are descriptions about the approach and the data structures used in the comments. 

A discussion about my summer work can be found in http://ramcha24.github.io/

My concern is with the API interfacing in linprog.jl . 
Please let me know if the current changes to linprog.jl are the right way to proceed.

### CODE REVIEW

File: src/HighLevelInterface/linprog.jl [hunk 1]
```diff
@@ -88,8 +86,29 @@ function linprog(c::InputVector, A::AbstractMatrix, rowlb::InputVector, rowub::I
     end
 end
 
+function linprog(c::InputVector, A::AbstractMatrix, rowlb::InputVector, rowub::InputVector, lb::InputVector, ub::InputVector, solver::AbstractMathProgSolver = MathProgBase.defaultLPsolver, presolve::Bool)
```
Comment (madeleineudell) [inline src/HighLevelInterface/linprog.jl]: I'd suggest a different factoring of the code: 1) adding `presolve::Bool=false` as an extra argument to the `linprog` above, 2) putting the code below into another function called `presolve_linprog` 3) having `linprog` call `presolve_linprog` if `presolve` is true.

File: src/HighLevelInterface/presolve.jl [hunk 2]
```diff
@@ -0,0 +1,905 @@
+#workspace()
```
Comment (madeleineudell) [inline src/HighLevelInterface/presolve.jl]: delete this line

File: src/HighLevelInterface/presolve.jl [hunk 3]
```diff
@@ -0,0 +1,905 @@
...
+
+`
+This is an implementation of LP Presolving based on ideas from Andersen & Andersen paper - http://www.turma-aguia.com/davi/doc/Andersen.pdf
```
Comment (madeleineudell) [inline src/HighLevelInterface/presolve.jl]: Doesn't this all need to be commented?

**Comment by mlubin** [comment]
@ramcha24, good to see your progress here. However, I'm not really sure this should all go into the MathProgBase package itself. ``linprog`` is not the main pathway for solving LPs (neither JuMP nor Convex.jl use it), so modifying ``linprog`` doesn't expose your code to JuMP/Convex.jl users.

The easiest way to make the presolve accessible to JuMP/Convex.jl users without modifying JuMP/Convex.jl is to make a fake "solver" that users can wrap around their actual LP solver, e.g., for JuMP, users would write something like
```
using JuMP, RamchaPresolve
m = Model(solver=Presolver(ClpSolver()))
```
and then the ``Presolver`` would implement the MathProgBase linear solver interface and transparently presolve the problem, pass it to the real solver, and then un-presolve the result. In this supposed RamchaPresolve package you could also implement a version of ``linprog`` that has an option for presolve, or people can use the MathProgBase ``linprog`` and just pass in ``Presolver(ClpSolver())`` as the solver. Once this is available for people to experiment with we might shake out the bugs and determine in which situations it's useful to presolve or not. *After* we understand this we could think about integrating it more deeply.

File: src/HighLevelInterface/presolve.jl (hunk 3 above)
Comment (ramcha24) [inline src/HighLevelInterface/presolve.jl]: Yes, it makes more sense to make them the docstring. Will do that.

**Comment by ramcha24** [comment]
@mlubin  @madeleineudell  , I understand the issue. So shall I make the fake Presolver? This idea also also means I can add Presolve support for conic problems and maybe even IP at a later stage. 

How do I go about making this ? If my understanding is right, this should be a higher layer of abstraction than MPB.

**Comment by mlubin** [comment]
@ramcha24, there are a number of examples of "fake" solvers that might do a bit of transformation and use another solver underneath:
https://github.com/JuliaOpt/MathProgBase.jl/blob/master/src/SolverInterface/conic_to_lpqp.jl
https://github.com/JuliaOpt/MathProgBase.jl/blob/master/src/SolverInterface/lpqp_to_conic.jl
https://github.com/JuliaOpt/MathProgBase.jl/blob/master/src/SolverInterface/nonlinear_to_lpqp.jl
https://github.com/mlubin/ConicNonlinearBridge.jl

Your solver should implement the "LinearQuadratic" interface to act as an LP solver, or if you prefer you could implement the "Conic" interface and only support the linear cones.

**Comment by ramcha24** [comment]
Thanks. I will try this.

**Comment by ramcha24** [comment]
@mlubin , you can look at my working repository for the MPB wrapper - https://github.com/ramcha24/Presolve 
There are one or two presolve routines to be added yet and we can start testing it after that's done.

**Comment by mlubin** [comment]
@ramcha24, the structure looks good. Could you add some usage examples with Convex.jl and JuMP to the README?

**Comment by mlubin** [comment]
This work is valuable but falls outside of the scope of what we would like to include in MathProgBase.jl. Closing this PR.

**Pull request closed without merging**